### PR TITLE
babel5 async/await

### PR DESCRIPTION
**Pull request by rom1504**
completely untested

### CODE REVIEW

**Comment by rom1504** [comment]
maybe do getter.get -> read (bind it and everything)

**Comment by rom1504** [comment]
looks pretty good. I'm going to make it work next.

**Comment by rom1504** [comment]
I think using the same idea I can make sizeof go away and writeX take a write. That will also let the serializer decide when he wants to send the data, in the middle of a packet if that's fitting

**Comment by roblabla** [comment]
This looks really good. Concerning write, you can send part of the packet to the compress stream.

**Comment by rom1504** [comment]
I fixed the tests and the benchmark.

But sadly it makes the benchmark 10 times slower. I'm not sure what to think about that...

**Comment by roblabla** [comment]
https://github.com/rvagg/bl maybe we could use this instead of concatting the buffer. 

Also concerning async/await, a good way to speed stuff up would be to use the `babel-plugin-transform-async-to-module-method` and bluebird for the async/await transform, instead of relying on generators. Generators are pretty damn slow right now. If we could somehow make babel transform await in the "natural-style" promise.then(), we'd probably see a perf bump.

**Comment by rom1504** [comment]
bl would indeed help, I'll try using that.

Is there any information anywhere about `babel-plugin-transform-async-to-module-method` ? I'm not sure how to use that

**Comment by rom1504** [comment]
https://kpdecker.github.io/six-speed/

generators, hence async/await are really slow.

I'll try rewriting with .then to see if that really improves perf a lot

**Comment by rom1504** [comment]
consider using https://www.npmjs.com/package/nodent which claims better async/await perfs (it doesn't use generators)

**Comment by rom1504** [comment]
Tentatives with circle ci times (10000 iteration + parser)
| Description | Time |
| --- | --- |
| master | 372ms |
| babel standard async/await | 6523ms |
| bluebird promises | 1078ms |
| babel promises | about 2,3 times bluebird promises |
| fast-async async/await | 4162ms |
| fast-async async/await with bluebird promises | 2015ms |

**Comment by rom1504** [comment]
Remaining possibilities:
* https://github.com/roblabla/ProtoDef/pull/39 babel6 + recent fast-async, might be better but can't do it before babel6 transform runtime get fixed
* https://github.com/roblabla/ProtoDef/pull/42 babel6 + bluebird coroutine, might be better but can't do it before babel6 transform runtime get fixed
* any other way to make async/await as fast as promises ?
* callback ? (just look quickly to see how much of a pain that would be)

Currently I think only the bluebird promises possibility is reasonably fast enough.

**Comment by roblabla** [comment]
Actually one last possibility. Nodent has a "Pure ES5 mode" that is supposed to be much faster. Might want to look at that ?

**Comment by rom1504** [comment]
Yeah I might try, but that probably means forking the plugin https://github.com/MatAtBread/fast-async/issues/1#issuecomment-161702769 , I might try though

**Comment by rom1504** [comment]
Alright, tried it, basing myself on https://github.com/roblabla/ProtoDef/pull/39
The problem is using the es5 transformation means you cannot use the Promise keyword, and I do need Promise there https://github.com/roblabla/ProtoDef/blob/babel6/src/serializer.js#L45
I guess it might be possible to write manually their es5 transformation instead of the Promise keyword but I'd better not go there.

**Comment by rom1504** [comment]
So that leaves 2 reasonable choices:
* [bluebird promises](https://github.com/roblabla/ProtoDef/pull/41) : best speed, style is okay if not as natural for some stuff (bitfields for example). Using only standard interfaces
* [fast-async babel6 bluebird](https://github.com/roblabla/ProtoDef/pull/39) : speed is okay, style is good, but it requires fitting stuff to the async/await particularities of nodent which I'm not sure we want to do (see [that](https://github.com/MatAtBread/fast-async/issues/1#issuecomment-161938724) )

Basically, I think using async/await heavily is too soon, there are still performance problems or interface problems. Going with the promise option is probably the safe bet for now. That is until the async/await get fixed in babel/node and then it will be easy to switch to it if we want to because promises are compatible with async/await.

**Comment by rom1504** [comment]
so I'm going to go with https://github.com/roblabla/ProtoDef/pull/41 , and I'm doing https://github.com/roblabla/ProtoDef/pull/43 next to try to improve the write interface and to check the parser actually handle split packets (which it doesn't right now apparently)

**Comment by rom1504** [comment]
closing in favor of #43
